packages/firestore/src/model/proto_values.ts

```diff
@@ -417,6 +417,63 @@ function canonifyArray(arrayValue: api.ArrayValue) : string {
   return result + ']';
 }
 
+/**
+ * Returns an approximate (and wildly inaccurate) in-memory size for the field
+ * value.
+ *
+ * The memory size takes into account only the actual user data as it resides
+ * in memory and ignores object overhead.
+ */
+export function estimateByteSize(value: api.Value): number {
+  if ('nullValue' in value) {
+    return 4;
+  } else if ('booleanValue' in value) {
+    return 4;
+  } else if ('integerValue' in value) {
+    return 8;
+  } else if ('doubleValue' in value) {
+    return 8;
+  } else if ('timestampValue' in value) {
+    // TODO(mrschmidt: Add ServerTimestamp support
+    // Timestamps are made up of two distinct numbers (seconds + nanoseconds)
+    return 16;
+  } else if ('stringValue' in value) {
+    // See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures:
+    // "JavaScript's String type is [...] a set of elements of 16-bit unsigned
+    // integer values"
+    return value.stringValue!.length * 2;
+  } else if ('bytesValue' in value) {
+    return normalizeByteString(value.bytesValue!).approximateByteSize();
+  } else if ('referenceValue' in value) {
+    return value.referenceValue!.length;
+  } else if ('geoPointValue' in value) {
+    // GeoPoints are made up of two distinct numbers (latitude + longitude)
+    return 16;
+  } else if ('arrayValue' in value) {
+    return estimateArrayByteSize(value.arrayValue!);
+  } else if ('mapValue' in value) {
+    return estimateMapByteSize(value.mapValue!);
+  } else {
+    return fail('Invalid value type: ' + JSON.stringify(value));
+  }
+}
+
+function estimateMapByteSize(mapValue: api.MapValue): number {
+  let size = 0;
+  forEach(mapValue.fields || {}, (key, val) => {
+    size += key.length + estimateByteSize(val);
+  });
+  return size;
+}
+
+function estimateArrayByteSize(arrayValue: api.ArrayValue): number {
+  let size = 0;
+  for (const value of arrayValue.values || []) {
+    size += estimateByteSize(value);
```

Comment:
Optional: this looks like `reduce`, though it doesn't look like it would change things substantially.

Reply:
I like it. Rewritten.